typescript/baseDAO/src/api.ts

```diff
@@ -0,0 +1,230 @@
+import { Set_quorum_threshold } from './generated/Set_quorum_threshold';
+import { Set_voting_period } from './generated/Set_voting_period';
+import { Transfer } from './generated/Transfer';
+import { Transfer_contract_tokens } from './generated/Transfer_contract_tokens';
+import { Transfer_ownership } from './generated/Transfer_ownership';
+import { Update_operators } from './generated/Update_operators';
+import { Vote } from './generated/Vote';
+
+function println(s: string) {
+  console.log(s);
+}
+
+type Callback = (Contract) => ContractMethod<ContractProvider>
+
+type ExtractMapKey<MapType> = MapType extends Map<infer R, infer V> ? R : null;
+type ExtractMapValue<MapType> = MapType extends Map<infer R, infer V> ? V : null;
+
+const unit = ["Unit"]; //https://github.com/ecadlabs/taquito/issues/526
+
+/**
+ * Wraps BaseDAO Contract methods
+ */
+export class BaseDAOContract {
+  private contractAddr: string;
+  private nodeAddr: string;
+  private senderSk: string;
+  private contract: undefined | Promise<ContractAbstraction<ContractProvider>>;
+
+  public debug: boolean;
+
+  public lastOperationHash: undefined | string;
+
+  constructor(nodeAddr: string, senderSk: string, contractAddr: string) {
+    this.contractAddr = contractAddr;
+    this.nodeAddr = nodeAddr;
+    this.senderSk = senderSk;
+    this.debug = false;
+  }
+
+  private initContract(): Promise<ContractAbstraction<ContractProvider>> {
+    if (this.contract === undefined) {
+      const Tezos = new TezosToolkit(this.nodeAddr);
+      Tezos.setProvider({ signer: new InMemorySigner(this.senderSk) });
+      this.contract = Tezos.contract.at(this.contractAddr);
+      return this.contract;
+    } else return this.contract;
+  }
+
+  public inspectParameter(): void {
+    this.initContract().then(contract => {
+      const methods = contract.parameterSchema.ExtractSignatures();
+      println(JSON.stringify(methods, null, 2));
+      });
+  }
+
+  private withContract(callback: Callback): Promise<string | void> {
+
+    // Initialize contract if it hasn't been initialized
+    return this.initContract().then(callback)
+        .then((method:ContractMethod<ContractProvider>) => {
+          if (this.debug) {
+            let tp = method.toTransferParams();
+            println(JSON.stringify(tp, null, 2));
+          }
+          return method.send();
+        })
+        .then((op: any): string => {
+          println(`Waiting for ${op.hash} to be confirmed...`);
+          return op.confirmation(1).then(() => {
+            this.lastOperationHash = op.hash;
+            return this;
+          });
+        })
+        //.catch((error) => println(`Error: ${JSON.stringify(error, null, 2)}`));
+        .catch((error) => println(`Error: ${error.toString()}`));
+  }
+
+  // entrypoint methods
+  balance_of(arg: Balance_of): Promise<string|void> {
+    return this.withContract((contract) => {
+      return contract.methods.balance_of(arg.requests, arg.callback);
+      });
```

Comment:
Nitpick: Just so this module is a bit smaller, when you have something like `(a) => { return b; }`, you may replace it with `a => b`.

Reply:
Right. Have made this change where possible.